File: iosApp/Modules/Domain/Login/LoginUseCase.swift
import Foundation
import API
import KeychainSwift
import FacebookLogin
import GoogleSignIn
import AuthenticationServices

enum LoginError: Error {
	case facebookAccessTokenMissing
}

public class LoginUseCase: ObservableObject {
	//TODO: update this with user info
	@Published public var userIsLoggedIn: Bool = false
	private let api: LoginAPI
	
	private var appleSignInID: String?
	
	//TODO: make actor
	public static let shared = LoginUseCase()
	
	private init(api: LoginAPI = LoginAPI()) {
		self.api = api
		api.$userIsLoggedIn.assign(to: &$userIsLoggedIn)
	}
		
	public func loginWithFacebook(result: LoginManagerLoginResult) async throws {
		do {
			guard let token = result.token?.tokenString else {
				throw LoginError.facebookAccessTokenMissing
			}
			try await api.login(method: .facebook(accessToken: token))
		} catch {
			Task { @MainActor in
				FacebookLogin.LoginManager().logOut()
			}
			throw error
		}
	}
	
	public func loginWithGoogle(result: GIDSignInResult) async throws {
		let token = result.user.accessToken.tokenString
		try await api.login(method: .google(accessToken: token))
	}
	
	public func login(email: String, password: String) async throws {
		try await api.login(method: .email(email: email, password: password))
	}
		
	public func logOut() {
		api.logOut()
		FacebookLogin.LoginManager().logOut()
		GIDSignIn.sharedInstance.signOut()
		signOutOfApple()
		UserManager.shared.currentUser = nil
	}
	
	public func requestEmailVerificationCode(for email: String) async throws {
		try await api.requestEmailVerificationCode(for: email)
	}
	
	public func loginWithApple(result: ASAuthorization) async throws {
		let token = String(data: (result.credential as! ASAuthorizationAppleIDCredential).identityToken!, encoding: .utf8)!
		try await api.login(method: .apple(idToken: token))
	}
	
	private func signOutOfApple() {
		//TODO:
	}
}

File: shared/src/commonMain/kotlin/io.newm.shared/usecases/LoginUseCase.kt
package io.newm.shared.usecases

import io.newm.shared.login.repository.KMMException
import io.newm.shared.login.repository.LogInRepository
import io.newm.shared.login.repository.OAuthData
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import kotlin.coroutines.cancellation.CancellationException

interface LoginUseCase {
    @Throws(KMMException::class, CancellationException::class)
    suspend fun logIn(email: String, password: String)

    @Throws(KMMException::class, CancellationException::class)
    suspend fun logInWithGoogle(idToken: String)

    @Throws(KMMException::class, CancellationException::class)
    suspend fun logInWithFacebook(accessToken: String)

    @Throws(KMMException::class, CancellationException::class)
    suspend fun logInWithLinkedIn(accessToken: String)

    @Throws(KMMException::class, CancellationException::class)
    suspend fun logInWithApple(idToken: String)

    fun logOut()
    val userIsLoggedIn: Boolean
}

internal class LoginUseCaseImpl(private val repository: LogInRepository) : LoginUseCase {
    @Throws(KMMException::class, CancellationException::class)
    override suspend fun logIn(email: String, password: String) {
        return repository.logIn(email = email, password = password)
    }

    @Throws(KMMException::class, CancellationException::class)
    override suspend fun logInWithGoogle(idToken: String) {
        return repository.oAuthLogin(OAuthData.Google(idToken))
    }

    @Throws(KMMException::class, CancellationException::class)
    override suspend fun logInWithFacebook(accessToken: String) {
        return repository.oAuthLogin(OAuthData.Facebook(accessToken))
    }

    @Throws(KMMException::class, CancellationException::class)
    override suspend fun logInWithLinkedIn(accessToken: String) {
        return repository.oAuthLogin(OAuthData.LinkedIn(accessToken))
    }

    @Throws(KMMException::class, CancellationException::class)
    override suspend fun logInWithApple(idToken: String) {
        return repository.oAuthLogin(OAuthData.Apple(idToken))
    }

    override fun logOut() {
        repository.logOut()
    }

    override val userIsLoggedIn: Boolean
        get() = repository.userIsLoggedIn()
}

class LoginUseCaseProvider(): KoinComponent {
    private val loginUseCase: LoginUseCase by inject()

    fun get(): LoginUseCase {
        return this.loginUseCase
    }
}